src/main/java/net/neoforged/neoforge/transfer/resources/ItemResource.java

```diff
@@ -0,0 +1,303 @@
+
+    /**
+     * Codec for an item resource. Same format as {@link #CODEC}, and also accepts empty resources.
+     */
+    public static final Codec<ItemResource> OPTIONAL_CODEC = Codec.lazyInitialized(() -> ExtraCodecs.optionalEmptyMap(CODEC).xmap(ItemResource::fromOptional, ItemResource::asOptional));
+
+    /**
+     * A codec for a {@code ResourceStack<ItemResource>} serializing the resource and the amount. Can accept empty resources.
+     */
+    public static final Codec<ResourceStack<ItemResource>> RESOURCE_STACK_CODEC = Codec.lazyInitialized(() -> ResourceStack.codec(OPTIONAL_CODEC, ItemResource::withAmount));
+
+    /**
+     * Stream codec for an item resource. Accepts empty resources.
+     */
+    public static final StreamCodec<RegistryFriendlyByteBuf, ItemResource> STREAM_CODEC = StreamCodec.composite(
+            ByteBufCodecs.holderRegistry(Registries.ITEM), ItemResource::getHolder,
+            DataComponentPatch.STREAM_CODEC, ItemResource::getComponentsPatch,
+            ItemResource::of);
+
+    /**
+     * Stream codec for a resource stack backed by an ItemResource. Accepts empty resources.
+     */
+    public static final StreamCodec<RegistryFriendlyByteBuf, ResourceStack<ItemResource>> RESOURCE_STACK_STREAM_CODEC = ResourceStack.streamCodec(ItemResource.STREAM_CODEC, ItemResource::withAmount);
+
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    public static ItemResource fromOptional(Optional<ItemResource> optional) {
+        return optional.orElse(ItemResource.EMPTY);
+    }
+
+    public Optional<ItemResource> asOptional() {
+        return isEmpty() ? Optional.empty() : Optional.of(this);
+    }
+
+    /**
+     * A helper method to quickly construct an {@link ItemStack} from a ResourceStack
+     *
+     * @param resourceStack The resource stack with the fluid resource and amount
+     * @return A new item stack with the same size as the resourceStack.
+     */
+    public static ItemStack itemStackOf(ResourceStack<ItemResource> resourceStack) {
+        return resourceStack.resource().toStack(resourceStack.amount());
+    }
+
+    /**
+     * This is used only for registry, you should not use this method!
+     */
+    @ApiStatus.Internal
+    public static ItemResource invalidateDefault(ItemLike item) {
+        if (item.asItem() == Items.AIR) return EMPTY;
+        return new ItemResource(new ItemStack(item));
+    }
+
+    public static ItemResource of(ItemStack itemStack) {
+        if (itemStack.isEmpty()) return EMPTY;
+
+        if (itemStack.isComponentsPatchEmpty())
+            return itemStack.getItem().getDefaultResource();
+
+        return new ItemResource(itemStack.copyWithCount(1));
+    }
+
+    /**
+     * <strong>Note:</strong> This cannot be called before your item is registered
+     */
+    public static ItemResource of(ItemLike item) {
+        if (item.asItem() == Items.AIR) return EMPTY;
+        return item.asItem().getDefaultResource();
```

Comment:
Can we add runtime assertions for this? 

What happens currently when you try to create an `ItemResource` for unregisted items?

Reply:
That's fine then. The documentation just needs to be updated to reflect that correctly.